library/src/scala/Tuple.scala

```diff
@@ -88,30 +90,24 @@ sealed trait Tuple extends Product:
   inline def reverseOnto[This >: this.type <: Tuple, Acc <: Tuple](acc: Acc): ReverseOnto[This, Acc] =
     (this.reverse ++ acc).asInstanceOf[ReverseOnto[This, Acc]]
 
-  /** A tuple consisting of all elements of this tuple that have types
-   *  for which the given type level predicate `P` reduces to the literal
-   *  constant `true`.
-   */
-  inline def filter[This >: this.type <: Tuple, P[_] <: Boolean]: Filter[This, P] =
-    val toInclude = constValueTuple[IndicesWhere[This, P]].toArray
-    val arr = new Array[Object](toInclude.length)
-    for i <- 0 until toInclude.length do
-      arr(i) = this.productElement(toInclude(i).asInstanceOf[Int]).asInstanceOf[Object]
+  /** A tuple consisting of all elements of this tuple that satisfy the predicate `p`. */
+  inline def filter[This >: this.type <: Tuple, P[_ <: Union[This]] <: Boolean]
+                   (p: (x: Union[This]) => P[x.type]): Filter[This, P] =
```

Comment:
question - should this be a poly function (like map), or continue as is? I guess with map the intention is to box a value and not inspect it, but with filter you actually need to read the value

Reply:
Imo `def map` should not have been poly fun either

#19600 has a version of map with dependant function type if you're interested it what the diffs look like

Reply:
If `filter` only worked on types, why should it be a method? it seems more like something you would accomplish with `constValueTuple[Filter[This, P]]`